Feature altimeter

sishamilton opened the pull request:
**Description**
The goal of this pull request is to introduce the Altimeter node.

The altimeter node is a sensor that gives altitude information.

**Related Issues**
This pull-request fixes issue #2767 

**Tasks**
Add the list of tasks of this PR.

**Webots Core**
  - [x] Add Altimeter.wrl in webots/resources/nodes/.
  - [x] Add a .png image of size 128x128 pixels in webots/resources/nodes/icons.
  - [x] Add WbAltimeter.hpp and WbAltimeter.cpp in webots/src/webots/nodes/.
  - [x] Add entry in webots/src/webots/Makefile.
  - [x] Add an enum entry in webots/include/controller/c/webots/nodes.h and webots/include/controller/cpp/webots/Node.hpp.
  - [x] Add the node name in wb_node_get_name() function in webots/src/controller/c/node.c.
  - [x] Add a gen_const() entry in webots/src/controller/matlab/mgenerate.py and run it.
  - [x] Add the node in src/webots/nodes/utils/WbConcreteNodeFactory.cpp.
  - [x] Add the node in src/webots/nodes/utils/WbNodeUtilities.cpp.
  - [x] Add the node in src/webots/core/WbLanguage.cpp.

**Changes Required for New Device**
  - [x] Add in webots/lib/controller/matlab/launcher.m and webots/lib/controller/matlab/allincludes.h the entries corresponding to the device node.
  - [x] Add the C++ controller class in webots/resources/templates/cpp.vcxproj and cpp.vcxproj.filter.
  - [x] Declare and define getAltimeter() and CreateAltimeter() in Robot.hpp and Robot.cpp and include in Robot.hpp.
  - [x] Add extern void wb_altimeter_init() in device.c.
  - [x] Add branch to switch in device.c::wb_device_init().
  - [x] Add RosAltimeter.cpp and RosAltimeter.hpp
  - [x] Add RosAltimeter to Ros.cpp constructor in webots/projects/default/controller/ros/.
  - [x] Add a demo featuring this new device in webots/projects/samples/devices/ and add it in the webots/projects/guided_tour.txt.

**New API Functions**:
  - [x] Add altimeter.h in webots/include/controller/c/webots/ and Altimeter.hpp in webots/include/controller/cpp/webots/.
  - [x] Add altimeter.c in webots/src/controller/c/ and Altimeter.cpp in webots/src/controller/cpp/.
  - [x] Add wb_altimeter_enable( ), wb_altimeter_disable( ), wb_altimeter_get_sampling_period( ), and wb_altimeter_get_value( ) API functions and equivalent C++ API functions.
  - [x] Add C API functions in webots/src/controller/c/Controller.def.
  - [x] Add API functions in /src/controller/java/ and src/controller/python/
  - [x] Create Altimeter MATLAB API with /src/controller/matlab/mgenerate.py
  - [x] Add API functions in /src/webots/core/WbLanguage.cpp
  - [x] Add RosAltimeter.hpp & .cpp in /projects/default/controllers/ros/
  - [x] Add API tests in /tests/api

**Documentation**
- [x] Update documentation including supervisor.md, nodes-and-keywords.md, and node-chart.md
- [x] Create altimeter.md file. Documentation will be found at: https://cyberbotics.com/doc/reference/altimeter?version=develop

**Screenshots**
If this pull-request includes any new robots/simulations/etc. add one or more screenshots of the result.

ad-daniel reviewed (commented):
Hi, thank you for this very nice contribution!
Overall it looks very good. As the PR is still in draft these are just some minor comments/suggestions. We'll make a full review when you think it's ready.

lukicdarkoo commented on projects/default/controllers/ros/RosAltimeter.cpp:
```diff
@@ -0,0 +1,41 @@
+  webots_ros::Float64Stamped value;
+  value.header.stamp = ros::Time::now();
+  value.header.frame_id = mRos->name() + '/' + RosDevice::fixedDeviceName();
+  value.altitude = mAltimeter->getValue();
```
```suggestion
  value.data = mAltimeter->getValue();
```
The ROS part is a bit tricky. Make sure you have ROS installed and before the `make` command make sure you have the ROS sourced in that terminal. That way the ROS controller will be compiled.

Note that later you will need to add a ROS test here:
https://github.com/cyberbotics/webots_ros/blob/master/src/complete_test.cpp

sishamilton replied:
Hello, I am not very familiar with Git submodules. To make changes to the webots_ros repo, should I fork it separately and open a pull request for my additions to complete_test.cpp, or is there a way for me to propose those changes from my fork of the main Webots repo?

omichel replied:
simply typing:
```
git submodule init
git submodule update
```
Should be sufficient to ensure your are up-to-date, you can check it with `git status`.

omichel replied:
> To make changes to the webots_ros repo, should I fork it separately and open a pull request for my additions to complete_test.cpp

Yes, that's the way to go.

lukicdarkoo commented on projects/default/controllers/ros/RosAltimeter.cpp:
```diff
@@ -0,0 +1,41 @@
+  value.header.frame_id = mRos->name() + '/' + RosDevice::fixedDeviceName();
+  value.altitude = mAltimeter->getValue();
+  publisher.publish(value);
+}
```
```suggestion
}

```
Always add a new line at the end of a file.

lukicdarkoo commented on tests/api/worlds/altimeter.wbt:
```diff
@@ -0,0 +1,58 @@
+#VRML_SIM R2021a utf8
+WorldInfo {
+  coordinateSystem "NUE"
```
```suggestion
```
We want to push ENU as a default coordinate system. Therefore, it would be nice that new tests use ENU (it is more important that a device works properly for ENU than for NUE)

lukicdarkoo commented on tests/api/worlds/altimeter.wbt:
```diff
@@ -0,0 +1,58 @@
+#VRML_SIM R2021a utf8
```
```suggestion
#VRML_SIM R2021b utf8
```
Make sure you are in the develop branch.

lukicdarkoo reviewed (changes requested):
Very nice!

I don't see any major issues, only a few small points.

ad-daniel commented:
Sorry my bad, looks like in some of the suggestions I made above it added a useless blank line. I just wanted to point out the issue/typo, the blank lines should not be there.

omichel commented:
@sishamilton: I just sent you an invite to become a member of the Webots Committers, which should allow you to launch the CI test and eventually merge this PR.

lukicdarkoo commented:
We fixed the Windows and Linux build bugs, so you can sync the feature branch with the develop branch to avoid relevant CI test failures

ad-daniel commented:
Good work so far, thank you.
When you believe you're finished, make sure to set it `ready for review` so that we can take a closer look at the PR.

sishamilton commented:
Sorry for the delay in completing this PR. I made the necessary changes in my webots_ros PR to introduce tests for the ROS Altimeter API ([#63](https://github.com/cyberbotics/webots_ros/pull/63)), so that should be ready to merge. Once that PR gets accepted, I can point my Webots branch to the correct version of the webots_ros submodule. After that I believe the only thing left to do for this pull request is to create a demo of the altimeter.

omichel commented on docs/reference/altimeter.md:
```diff
@@ -0,0 +1,134 @@
+
+- `resolution`: This field allows the user to define the resolution of the sensor. The resolution is the smallest change that it is able to measure.
+Setting this field to -1 (default) means that the sensor has an 'infinite' resolution (it can measure any infinitesimal change).
+This field accepts any value in the interval (0.0, inf).
```
```suggestion
Otherwise this field accepts any value in the interval [0.0, inf).

```

omichel commented on docs/reference/altimeter.md:
```diff
@@ -0,0 +1,134 @@
+
+The `wb_altimeter_get_sampling_period` function returns the period given into the `wb_altimeter_enable` function, or 0 if the device is disabled.
+
+---
```
```suggestion
```

omichel commented on include/controller/c/webots/altimeter.h:
```diff
@@ -0,0 +1,41 @@
+}
+#endif
+
+#endif /* WB_ALTIMETER_H */
```
```suggestion
#endif /* WB_ALTIMETER_H */

```

omichel commented on include/controller/cpp/webots/Altimeter.hpp:
```diff
@@ -0,0 +1,32 @@
+  };
+}  // namespace webots
+
+#endif  // ALTIMETER_HPP
```
```suggestion
#endif  // ALTIMETER_HPP

```

omichel commented on projects/default/controllers/ros/RosAltimeter.hpp:
```diff
@@ -0,0 +1,40 @@
+  Altimeter *mAltimeter;
+};
+
+#endif  // ROS_RANGE_FINDER_HPP
```
```suggestion
#endif  // ROS_RANGE_FINDER_HPP

```

omichel commented on projects/samples/devices/controllers/altimeter/altimeter.c:
```diff
@@ -0,0 +1,63 @@
+
+  wb_robot_cleanup();
+  return 0;
+}
```
```suggestion
}

```

omichel commented on resources/nodes/Altimeter.wrl:
```diff
@@ -0,0 +1,29 @@
+  field     SFFloat    rotationStep        0.261799387 # step size used by rotation manipulator
+  field     SFFloat    radarCrossSection   0.0         # radar cross section of this solid
+  field     MFColor    recognitionColors   []          # colors returned for this Solid by Cameras with a Recognition node
+  #fields specific to the GPS node:
```
```suggestion
  #fields specific to the Altimeter node:

```

omichel commented on resources/nodes/Altimeter.wrl:
```diff
@@ -0,0 +1,29 @@
+  # hidden fields
+  hiddenField SFVec3f linearVelocity       0 0 0   # (m/s) Solid's initial linear velocity
+  hiddenField SFVec3f angularVelocity      0 0 0   # (rad/s) Solid's initial angular velocity
+}
```
```suggestion
}

```

omichel commented on resources/templates/controllers/cpp.vcxproj:
```diff
@@ -121,6 +121,8 @@
   <ItemGroup>
     <ClCompile Include="$(WEBOTS_HOME)\src\controller\cpp\Accelerometer.cpp" />
+    <ClCompile Include="$(WEBOTS_HOME)\src\controller\cpp\Altimeter.cpp"/>
+
```
```suggestion
```

omichel commented on resources/templates/controllers/cpp.vcxproj.filter:
```diff
@@ -24,6 +24,10 @@
     </ClCompile>
+    <ClCompile Include="$WEBOTS_HOME)\src\controller\cpp\Altimeter.cpp">
+      <Filter>Source Files\cpp</Filter>
+
```
```suggestion
```

omichel commented on src/controller/cpp/Altimeter.cpp:
```diff
@@ -0,0 +1,38 @@
+
+double Altimeter::getValue() const {
+  return wb_altimeter_get_value(getTag());
+}
```
```suggestion
}

```

omichel commented on src/controller/java/controller.i:
```diff
@@ -838,6 +846,17 @@ namespace webots {
   }
 
+  protected Altimeter createAltimeter(String name) {
+      return new Altimeter(name);
```
```suggestion
    return new Altimeter(name);
```

omichel commented on src/webots/nodes/WbAltimeter.cpp:
```diff
@@ -0,0 +1,133 @@
+  // compute current altitude
+  double accuracy = mAccuracy->value();
+
+  WbVector3 reference = WbWorld::instance()->worldInfo()->gpsReference();
```
```suggestion
  const WbVector3 reference = WbWorld::instance()->worldInfo()->gpsReference();
```

omichel commented on src/webots/nodes/WbAltimeter.cpp:
```diff
@@ -0,0 +1,133 @@
+
+  WbVector3 reference = WbWorld::instance()->worldInfo()->gpsReference();
+  const QString &coordinateSystem = WbWorld::instance()->worldInfo()->coordinateSystem();
+  int upIndex = coordinateSystem.indexOf('U');
```
```suggestion
  const int upIndex = coordinateSystem.indexOf('U');
```

omichel commented on src/webots/nodes/WbAltimeter.cpp:
```diff
@@ -0,0 +1,133 @@
+void WbAltimeter::writeConfigure(QDataStream &stream) {
+  mSensor->connectToRobotSignal(robot());
+  addConfigureToStream(stream);
+}
```
```suggestion
}

```

omichel commented on src/webots/nodes/WbAltimeter.hpp:
```diff
@@ -0,0 +1,66 @@
+  void updateResolution();
+};
+
+#endif
```
```suggestion
#endif

```

omichel commented on tests/api/controllers/altimeter/.gitignore:
```diff
@@ -0,0 +1 @@
+/altimeter
```
```suggestion
/altimeter

```

omichel commented on tests/api/controllers/altimeter/Makefile:
```diff
@@ -0,0 +1,25 @@
+null :=
+space := $(null) $(null)
+WEBOTS_HOME_PATH=$(subst $(space),\ ,$(strip $(subst \,/,$(WEBOTS_HOME))))
+include $(WEBOTS_HOME_PATH)/resources/Makefile.include
```
```suggestion
include $(WEBOTS_HOME_PATH)/resources/Makefile.include

```

omichel commented on tests/api/controllers/altimeter/altimeter.c:
```diff
@@ -0,0 +1,38 @@
+
+  ts_send_success();
+  return EXIT_SUCCESS;
+}
```
```suggestion
}

```

omichel reviewed (changes requested):
Thank you for this great contribution!
I only have minor change requests.

The MATLAB functions should not be committed as they are automatically generated now. See new instructions at https://github.com/cyberbotics/webots/wiki/Adding-New-Node-and-API-Function#adding-a-new-api-function about MATLAB.

ad-daniel commented on docs/reference/altimeter.md:
```diff
@@ -0,0 +1,134 @@
+### Description
+
+The [Altimeter](#altimeter) node is used to model an altimeter sensor, which can obtain information about its height above the global reference plane from the controller program.
+
```
```suggestion

```

ad-daniel commented on lib/controller/matlab/wb_altimeter_disable.m:
```diff
@@ -0,0 +1,6 @@
+function wb_altimeter_disable(tag)
+% Usage: wb_altimeter_disable(tag)
+% Matlab API for Webots
+% Online documentation is available <a href="https://www.cyberbotics.com/doc/reference/altimeter">here</a>
+
+calllib('libController', 'wb_altimeter_disable', tag);
```
Given a very recent change, it is no longer necessary to push the `.m` files.
Same thing for the other 3.

ad-daniel reviewed (commented):
Just mentioned two minor changes I noticed.
Also, concerning the image `Altimeter.png`, you made and/or is it free to use? Or does it require attribution? A free option is preferable.
Additionally, the image should have a transparent background in case people use a dark template.

Thank you for your contribution!

sishamilton commented:
The Altimeter.png is free for commercial use without attribution. I went ahead and replaced the image with a version with a transparent background.

omichel commented:
Thank you. There are still a few things missing:

1. The MATLAB files were not removed as requested.
2. The const related suggestions were not taken into account.
3. The image of the altimeter is now rotated by 90 degrees.

Otherwise, everything looks good to me.

sishamilton commented:
Ok, I removed the Altimeter Matlab API files, accepted the const related changes, and provided a properly oriented icon.

omichel reviewed (approved):
Thank you.
I just added the missing entries in the `.gitignore` of the `lib/controller/matlab` folder.

ad-daniel reviewed (dismissed):
looks good to me, thank you!

lukicdarkoo reviewed (changes requested):
Looks good, but I have a few comments:

- The icon doesn't look very nice in the dark Webots version:
   ![image](https://user-images.githubusercontent.com/2135826/123637119-eefad800-d81d-11eb-91c7-72c6b87cec16.png)
- There is a missing `accuracy` parameter in the VRML prototype:
https://cyberbotics.com/doc/reference/altimeter?version=sishamilton:feature-altimeter
- There is a `noiseCorrelation` parameter that doesn't serve any purpose.

sishamilton commented:
@lukicdarkoo I removed the `noiseCorrelation` property from the documentation and Altimeter.wrl.

I also added a cleaner image for the Altimeter icon. It looks better with the dark theme.
![image](https://user-images.githubusercontent.com/69378550/123723060-75271680-d83e-11eb-960b-0d162fb8c56c.png)

omichel commented:
This branch was messed-up after attempting to resolve conflicts with the develop branch.
It is replace by #3241.
